### DISCUSSION TITLE
[Done] Support Tailwind 3.0 JIT

### DISCUSSION THREAD

**Discussion by filipenevola** [meteor/meteor#11804] [Feature Requests]
Hi, let's use this to discuss the better way to implement JIT with Meteor build tool.

**Comment by rijk** [comment]
Related: https://github.com/Meteor-Community-Packages/meteor-postcss/issues/51

**Comment by Saeeed-B** [comment]
@filipenevola @StorytellerCZ 
Undoubtedly, this feature should be added as soon as possible, the greater the need for it.
On the other hand, in version 3, we encountered many problems that we are not able to build an app.

**Comment by gabrilator** [comment]
This would be a game changer as I think Tailwind 3.0 offers many new improvements !

**Comment by ixam1** [comment]
I´d like to bump this feature again.

I am not sufficient with postcss but the problem seems to lie with postcss not registering file changes while in dev to trigger the css to be rebuilt. Currently you have to restart meteor to trigger the build again and thus the whole concept of Just-in-Time mode doesnt work. 

Maybe this has something to do with the following:

[Quote](https://v2.tailwindcss.com/docs/just-in-time-mode#troubleshooting)

> ### Styles don't update when saving content 
> As of Tailwind CSS v2.2+, the JIT engine depends on PostCSS’s directory dependency messages to register your content files as CSS build dependencies with your build tool. These are a fairly new addition to PostCSS (added in May 2021), and not all build tools have been updated to support them yet.

So maybe meteor postcss "just" needs to be updated to support directory dependency messages?

**Reply by wojtekgalaj** [reply]
Check out this solution. It works very well for me.

https://github.com/Meteor-Community-Packages/meteor-postcss/issues/51#issuecomment-997200148

**Reply by ixam1** [reply]
@wojtekgalaj Thanks for pointing me to that thread, didnt know it existed. 

Anyway isnt that solution basically the same as [this solution using the tailwind cli instead](https://forums.meteor.com/t/new-tailwind-css-jit-compiler/55592/12)? I do it like that currently and it works, but it causes double reloads. Change file -> meteor reloads + CLI rebuilds css -> meteor reloads

I can live with it, but with the already slow reload times I have with meteor, it really bothers me.

Or don´t you have double reloads with the postcss cli?

**Reply by wojtekgalaj** [reply]
I don't think I get double reload. I might be wrong and I can't check now but if I do it's seamless. Might just be that my project is simple.

**Reply by wojtekgalaj** [reply]
Just to be clear, I would love to not have to rely on this solution. It is a bit hacky and I think meteor should handle that. 

I'm too stupid to figure out how to fix it though.

**Reply by ixam1** [reply]
What are your hot reload times? Mine is a few seconds and doubling that gets really noticable. My application isnt huge but not small either. I would love to figure out why my reloads are so slow, but hadnt had the time and its relatively low priority... although if you would add all those seconds wasted, I´d imagine I would get quite a big sum.

> Just to be clear, I would love to not have to rely on this solution. It is a bit hacky and I think meteor should handle that.
> 
> I'm too stupid to figure out how to fix it though.

Yeah I´m in the same boat.

**Reply by wojtekgalaj** [reply]
So I don't have this double reload problem. 

Try following that linked solution. I slightly modified it on my project so instead of having two processes, one for meteor and another for postcss, I run them both in one process using concurrently.

 ```    
    "start": "concurrently \"npm run meteor\" \"npm run postcss\"",
    "meteor": "meteor run --exclude-archs web.browser.legacy",
    "postcss": "postcss client/main.css -o client/post.css --watch",
```

Other than that, everything is setup as in the linked solution.

I only get one rebuild from meteor. Hot reload times for me are around half a second eyeballing it. It's really ok.

**Comment by gabrilator** [comment]
Agreed! I know it must be a handful, but not having good support for Tailwind can be  a dealbreaker for many people!

**Comment by filipenevola** [comment]
Hi, I updated our community [minifier](https://github.com/Meteor-Community-Packages/meteor-postcss/pull/55/files) according to Tailwind docs so we have the set up working as expected with the limitation that this discussion tries to solve: Meteor is not generating the new CSS classes on-the-fly.

So I started to investigate why this is happening.

I have some news, testing what could be the root cause of TailwindCSS 3.0 JIT mode not working with Meteor I found out some things.

- postcss minifier is working fine when CSS files are changed, so you could make a change to your App.js adding a new Tailwind class (Meteor is not going to add this class for you automatically to your final CSS file) and if you also make any random change to any CSS file the postcss minifier will run again (and then Meteor will add this new class for you) and produce the new Tailwind class for you.
  - Not great, you don't want to change a random CSS file all the time to update your Tailwind generated classes but at least now we know that everything else is running fine, the challenge is to change the behavior in Meteor for it to also re-run postcss when other files rather than CSS files are changed.
- forcing the postcss minifier to watch other extensions is not possible as each extension should have only one modifier and we still need the minifier for JS files anyway.
- even when I force the CSS minifier to be called it was not doing anything, this was happening because of this [optimistic wrap](https://github.com/meteor/meteor/blob/54557867e8f68658fb9786b6eb36e96de9478c6f/tools/isobuild/bundler.js#L1900). As no CSS files is changed, the cache key is still valid so the function is not executed.
- removing this optimistic wrap from this function makes everything work fine. Of course, postcss is going to run after every change.

Based on this I would say the solution is to improve this cache key to include other files that should trigger changes in Tailwind classes, what comes to mind immediately is the new property in tailwind.config.js: `content`. So we should check our files and make sure we generate the CSS again when these files are changed.

So let's start with a naive approach, our app code is always minified in the `app/app.js` file so instead of reading config from tailwind let's see if this simple approach would work:

Change this call to `minifyCss`:

https://github.com/meteor/meteor/blob/54557867e8f68658fb9786b6eb36e96de9478c6f/tools/isobuild/bundler.js#L911

to:

```js
this.minifyCss(minifiersByExt.css, minifyMode, {filesToCacheKey: this.js.filter(file => file.url === '/app/app.js')});
```

I'm getting the actual js file, already minified in the step before this one, and only providing it as an additional file to generate the cache key.

In order for this new file to be used in the cache key I had to change the minifyCss function from:

https://github.com/meteor/meteor/blob/54557867e8f68658fb9786b6eb36e96de9478c6f/tools/isobuild/bundler.js#L1689-L1695

to:

```js
  minifyCss(minifierDef, minifyMode, minifyOptions = {}) {
    this.css = minifyCssFiles(this.css, {
      arch: this.arch,
      minifier: minifierDef,
      minifyMode,
      minifyOptions,
    });
  }
```

and the `minifyCssFiles` `makeCacheKey` call from:

https://github.com/meteor/meteor/blob/54557867e8f68658fb9786b6eb36e96de9478c6f/tools/isobuild/bundler.js#L1947-L1954

to:

```js
  makeCacheKey(files, { arch, minifier, minifyMode, minifyOptions }) {
    const cacheFiles = [...files, ...(minifyOptions.filesToCacheKey || [])];
    return defaultMakeCacheKey(
      minifier,
      arch,
      minifyMode,
      hashOfFiles(cacheFiles),
    );
  }
```

And voilà! It is generating my CSS file, with the new CSS classes from Tailwind when I change anything in my app client-side code.

As I said, this was just a naive approach to validate the solution. Now with other contributors' feedback, we can improve this solution to decide the final approach.

I believe reading from `content` of tailwind.config.js is not the best approach especially because we could have other things there rather than js files but that also means that we would probably need to modify the idea above a bit as I don't think we have .html at hand, for example.

@zodern @renanccastro and all, I would like to hear your suggestions for a final solution.

**Reply by zodern** [reply]
That's interesting that isobuild caches the output of the css minifier. It should probably be handled instead by the build plugin since it has more information on how the caching should work, similar to how build plugins for js minifiers and compilers handle it.

Tailwind and PostCSS provide a list of dependencies: https://v2.tailwindcss.com/docs/just-in-time-mode#styles-don-t-update-when-saving-content-files
I haven't looked at it too closely, but it should be possible to have the build plugin add those to the watch set in case they aren't already and have the build plugin use that information for cache keys.

**Reply by filipenevola** [reply]
> but it should be possible to have the build plugin add those to the watch set in case they aren't already and have the build plugin use that information for cache keys.

I don't think the problem is the watch set.

The problem is the cache used by the minifier caller as I showed above.

Tailwind now creates the classes on-the-fly so every JS and HTML change is a potential change for generate new CSS classes. Anything that uses CSS classes should force a new postcss run on changes. Before JIT this was not necessary as tailwind was producing all the classes first and only purging the unused ones later.

Also, other postcss utilities were only necessary on CSS files changes instead of CSS classes usage in any file extension.

I believe we should fix this cache with a simple solution and later we could elaborate a better one, if necessary, changing how minifier are called and this cache is created.

**Reply by filipenevola** [reply]
I'll try to include HTML files in this cache key as well and move forward with a simple solution.

This solution would be enabled only when a specific setting is enabled and tailwind is used so we wouldn't affect other usages of minifier plugins.

**Comment by zodern** [comment]
I opened two PR's that fix using Tailwind JIT:
- https://github.com/meteor/meteor/pull/11882
- https://github.com/Meteor-Community-Packages/meteor-postcss/pull/56

**Reply by filipenevola** [reply]
Hi @zodern I've tested your changes and they are working as expected for JS files but I believe changes in HTML files are not triggering a new postcss run. Have you checked this case?

I've tested using my meteor-template repository with these [changes](https://github.com/quavedev/meteor-template/commit/84690af3e813acc1e523c8a9d2923ac90e189791) and creating a symbolic link for the postcss package pointing to this [branch](https://github.com/Meteor-Community-Packages/meteor-postcss/pull/56#pullrequestreview-870567405) and running Meteor locally with this [branch](https://github.com/meteor/meteor/pull/11882#pullrequestreview-870567653).

Performed Tests:
- ✅ Add new usage for TailwindCSS classes in [Home.js](https://github.com/quavedev/meteor-template/blob/tailwind3/app/home/Home.js). The new class should appear in the final CSS automatically.  
- 🚫 Add new usage for TailwindCSS classes in [main.html](https://github.com/quavedev/meteor-template/blob/tailwind3/client/main.html). The new class should appear in the final CSS automatically, this is not happening, I have to restart the server to see this new class. (it was fixed in the last commit of Zodern on postcss package).
- ✅ Overwrite a new custom color in [tailwind.config.js](https://github.com/quavedev/meteor-template/blob/tailwind3/tailwind.config.js).  The new color should be reflected in the browser automatically.

**Reply by fredmaiaarantes** [reply]
Hey @filipenevola I looked at the [changes](https://github.com/quavedev/meteor-template/commit/84690af3e813acc1e523c8a9d2923ac90e189791) you made and if you change your `tailwind.config.js` file from this:

```
content: [
    './app/**/*.{js,jsx,ts,tsx}',
    './client/**/*.{js,jsx,ts,tsx}',
    './client/*.html',
  ],
```

To this:

```
content: [
  './app/**/*.{js,jsx,ts,tsx}',
  './client/**/*.{js,html,jsx,ts,tsx}',
],
```

This way it worked on the tests I did.

**Reply by zodern** [reply]
There is a bug in how it handles the glob for `./client/*.html`. If the glob has `**` someplace it works correctly. I am looking into it.

**Reply by zodern** [reply]
It should be fixed by https://github.com/Meteor-Community-Packages/meteor-postcss/pull/56/commits/6580189d469c09b7004713723175e8396bce8e0e

**Reply by fredmaiaarantes** [reply]
Thanks @zodern! Now, it worked for me using it in the same way as @filipenevola did.

**Reply by filipenevola** [reply]
Now it works for me as well. Great @zodern.

- ✅  Add new usage for TailwindCSS classes in [main.html](https://github.com/quavedev/meteor-template/blob/tailwind3/client/main.html). The new class should appear in the final CSS automatically.

**Comment by filipenevola** [comment]
Hi, I have good news, to provide core support to CSS Minifier with PostCSS we are incorporating the community package into the core as a new core package: `minifier-css-postcss`.

The latest beta of Meteor (2.6.1) supports Tailwind 3.x.

Read the migration guide linked in the PR:
https://github.com/meteor/meteor/pull/11886

Feedback is welcome in the same PR linked above.

**Reply by zodern** [reply]
This is good. Instead of having two minifiers, would there be any issue with merging `standard-minifier-css` and `minifier-css-postcss`? Some bundlers support postcss automatically when there is a postcss config. I've lately been thinking it would be nice if Meteor would do the same.

Where is the code for `minifier-css-postcss`?

**Reply by radekmie** [reply]
@zodern asked both of my questions. I see no point in having two minifiers, especially when there’s no drawback of using `minifier-css-postcss`. Even if there’s a small performance hit (is there any? it should be measured), a single codebase makes more sense.

I also was looking for the code and couldn’t find it.

**Reply by filipenevola** [reply]
The code is here https://github.com/meteor/meteor/pull/11886/commits/7c3e623a094e581754ed442c34b6a4caf648fba1

WebStorm somehow didn't send the files, but they are there now.

We are open to suggestions, but at first, I thought it would be better in terms of isolation as I don't think everybody uses `postcss`.

**Reply by radekmie** [reply]
> […] as I don't think everybody uses `postcss`.

The standard minifier already relies on it, thus I think it’s not a problem.

**Reply by zodern** [reply]
If the app doesn't have a postcss config, then the standard minifier would work the same as it does now.  The only concern I have with merging the two minifiers is if for some reason the app has a postcss config for another tool, but doesn't want Meteor to use postcss. This was being done by some of the work arounds for using tailwind 3 with Meteor (https://github.com/Meteor-Community-Packages/meteor-postcss/issues/51#issuecomment-997200148). Though now that Meteor support's all of postcss's features it shouldn't be needed, and if someone doesn't want Meteor to use the App's postcss config there is the option for the app to switch to a different css minifier.

**Reply by radekmie** [reply]
Maybe this could be configured in Meteor-specific settings in `package.json` then? Like, an override, including the option to ignore other settings.

**Reply by zodern** [reply]
If we need to make it configurable, we could add an option, or we could keep it as two separate minifiers. Though it seems unlikely we would need to, and other bundlers that automatically enable postcss don't make it configurable, such as vite.

**Reply by filipenevola** [reply]
I don't have strong opinions on this matter.

Having fewer packages, in general, is better. What are the cons of unifying these two packages if there is a way to disable handling postcss?

Also, is it a bad idea to start with two packages, see how it goes, and later decide if we should unify or not?

If we start with two packages, we can write the docs on when to use each one, and with time, we will find if we need to have two packages. I would say this is a better path.

**Reply by zodern** [reply]
> What are the cons of unifying these two packages if there is a way to disable handling postcss?

I can't think of any.

I'm fine with them being two packages, though I'm not sure what benefits it has over merging them.

**Reply by radekmie** [reply]
For me the benefit of merging is that there’s no fuss of choosing one over the other. It’s also easier to explain for people upgrading their projects, that the minifier can now handle PostCSS plugins and there’s no need to swap it.

Unifying them later is also fine, but it’ll force _some_ people to migrate twice - once now, and another time after the merge.

**Reply by filipenevola** [reply]
I'm ok unifying them if we don't see any cons to doing it.

@zodern, could you do this in the work that you have already planned for next week?

**Reply by Saeeed-B** [reply]
As a user, the addition of more packages will definitely bring me more confusion,
There is a fear of package management in node js. Especially when we have to install packages for very small tasks
As a developer, I prefer to work centrally.

**Reply by zodern** [reply]
Here is the PR that unifies the two packages: https://github.com/meteor/meteor/pull/11903

**Comment by filipenevola** [comment]
2.6.1-beta.3 is available.

Updated [migration guide](https://deploy-preview-11886--meteor-guide.netlify.app/2.6.1-migration.html)

**Reply by mart-jansink** [reply]
Great work @filipenevola! However, I encountered an issue (https://github.com/meteor/meteor/issues/11940) while using packages from the local `packages` directory that's preventing startup. I'll try to remedy it myself as well of course.

**Reply by fredmaiaarantes** [reply]
Tailwind 3 support is coming on Meteor version 2.7.

- [Migration Guide Preview](https://deploy-preview-11916--meteor-guide.netlify.app/2.7-migration.html)
- PR: https://github.com/meteor/meteor/pull/11916

**Comment by topherfangio** [comment]
Hi all, I'm just playing around with Meteor again (after a ~5 year hiatus) and am looking for the best way to get started with Tailwind inside Meteor.

I looked at the migration guide listed above, but it doesn't really spell out exactly what I should do.

I was hoping for something like `meteor add tailwindcss` and it would automagically work, but I guess we're not quite there yet 😄 

So should I be looking to use the Tailwind CLI? Or is there a better approach?

Thanks!

**Reply by chrisallenmoore** [reply]
@topherfangio, [Meteor release 2.6.1](https://deploy-preview-11886--meteor-guide.netlify.app/2.6.1-migration.html#tailwind-css) has a little information on how to use Tailwind CSS 3.X. Since you're starting a fresh project, use the [Tailwind CLI](https://tailwindcss.com/docs/installation) instructions.

Here's an example of what you need to know for content paths in your `tailwind.config.js` file:


```
module.exports = {
// Example content paths
  content: ['./client/**/*.{html,js,svelte,ts}', './imports/**/*.{html,js,svelte,ts}'],
  theme: {
    extend: {
      //
      }
    },
    fontFamily: {
    }
  },
  plugins: [],
}
```


I use `svelte` in my projects, but feel free to replace it with whatever framework file endings you need.

**Comment by fredmaiaarantes** [comment]
Tailwind 3 support is coming on Meteor version 2.7.

- [Migration Guide Preview](https://deploy-preview-11916--meteor-guide.netlify.app/2.7-migration.html)
- PR: https://github.com/meteor/meteor/pull/11916

**Comment by fredmaiaarantes** [comment]
Meteor 2.7 is already out, with Tailwind 3 support, in a few days we will recommend it and announce it on our channels (blog, social, etc).

The docs are now up to date and you can see the [changelog](https://docs.meteor.com/changelog.html) and follow the [migration guide](https://guide.meteor.com/2.7-migration.html) to start using it.
Let us know if any issues are found.
